### PR TITLE
[CF-971] BC5 postReceipt updates

### PR DESCRIPTION
**Pull request by beylmk**
### Checklist
- [will be in separate PR] If applicable, unit tests
- [n/a, will follow up for hybrids once api is finalized] If applicable, create follow-up issues for `purchases-ios` and hybrids

### Motivation

In BillingClient5, we do not receive price information in the `Purchase` object returned when a purchase completes. Instead, we need to store which `PurchaseOption` the purchase was initiated for, pass the associated `PricingPhase`s to our backend, and let our backend handle the price reporting.

For restored purchases, we are not in control of the purchase flow and therefore don't know the `PurchaseOption` which triggered the purchase. In those cases, our backend will have to use the Google APIs to deduce the price associated with the Purchase.

Resolves https://revenuecats.atlassian.net/browse/CF-887

### Description

Store the `purchaseOptionId` associated with the `Purchase` and the associated purchase callback, similar to how we previously stored the `sku`. Then, when a purchase completes, pass the `PricingPhase`s of that `PurchaseOption` to the backend.

I also removed a lot of the fields we no longer need to post to the backend.

I ran through a purchase flow in the sample app and confirmed that postReceipt returned 200. Unit tests will be a in a separate PR, and end-to-end testing in the bug bash and with the testing restores ticket.

### CODE REVIEW

File: common/src/main/java/com/revenuecat/purchases/common/Backend.kt [hunk 1]
```diff
@@ -188,14 +189,14 @@ class Backend(
             "currency" to receiptInfo.currency,
             "attributes" to subscriberAttributes.takeUnless { it.isEmpty() },
             "normal_duration" to receiptInfo.duration,
-            "intro_duration" to receiptInfo.introDuration,
```
Comment (beylmk) [inline common/src/main/java/com/revenuecat/purchases/common/Backend.kt:191]: we still have to figure out how to pass these values in amazon...

amazon offers work the way google used to -- they are automatically applied. so i don't think we can make PurchaseOptions with PricingPhases for those. 

either way, i'm not sure we have that info available to us. Product.toStoreProduct for amazon just passed null introPrice/freeTrial fields. @vegaro can you confirm this?

File: common/src/main/java/com/revenuecat/purchases/common/Backend.kt [hunk 2]
```diff
@@ -188,14 +189,14 @@ class Backend(
...
-            "trial_duration" to receiptInfo.trialDuration,
-            "store_user_id" to storeAppUserID
+            "store_user_id" to storeAppUserID,
+            "pricing_phases" to receiptInfo.pricingPhases
         ).filterValues { value -> value != null }
 
-        val extraHeaders = /*receiptInfo.storeProduct?.price?.let { priceString ->
+        // TODO BC5 is price_string still expected? what to pass?
```
Comment (beylmk) [inline common/src/main/java/com/revenuecat/purchases/common/Backend.kt:196]: @dalshamaa @swehner thoughts on the old price_string header?

File: common/src/main/java/com/revenuecat/purchases/common/BillingAbstract.kt [hunk 3]
```diff
@@ -96,6 +96,13 @@ abstract class BillingAbstract {
      * the correct product ID the RevenueCat backend expects for a specific purchase.
      * Google doesn't need normalization so we return the productID by default
      */
+
+    // todo maddie
+    // this is my understanding, could you confirm?:
```
Comment (beylmk) [inline common/src/main/java/com/revenuecat/purchases/common/BillingAbstract.kt:101]: semi unrelated to this PR, just me trying to understand...

File: common/src/test/java/com/revenuecat/purchases/common/BackendTest.kt [hunk 4]
```diff
@@ -64,18 +64,24 @@ class BackendTest {
         mockClient
     )
     private val appUserID = "jerry"
+    private val storeProductNoOffers = stubStoreProduct("productID")
```
Comment (beylmk) [inline common/src/test/java/com/revenuecat/purchases/common/BackendTest.kt:67]: hold off on reviewing the tests please, still working on them

File: feature/google/src/main/java/com/revenuecat/purchases/google/BillingWrapper.kt [hunk 5]
```diff
@@ -72,6 +72,7 @@ class BillingWrapper(
 
     private val productTypes = mutableMapOf<String, ProductType>()
     private val presentedOfferingsByProductIdentifier = mutableMapOf<String, String?>()
+    private val purchaseOptionSelectedByProductIdentifier = mutableMapOf<String, String?>()
```
Comment (beylmk) [inline feature/google/src/main/java/com/revenuecat/purchases/google/BillingWrapper.kt:75]: perhaps these two maps could instead be a map<String, Pair<String?, String?>>

File: common/src/main/java/com/revenuecat/purchases/common/Backend.kt (hunk 2 above)
Comment (swehner) [reply common/src/main/java/com/revenuecat/purchases/common/Backend.kt:196]: I couldn't find any reference to this in the backend code...

File: common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt [hunk 6]
```diff
@@ -1,18 +1,22 @@
...
+import com.revenuecat.purchases.models.PricingPhase
 import com.revenuecat.purchases.models.StoreProduct
 
 class ReceiptInfo(
     val productIDs: List<String>,
     val offeringIdentifier: String? = null,
-    val storeProduct: StoreProduct? = null,
-    val price: Double? = 1.0, // TODOBC5 storeProduct?.priceAmountMicros?.div(MICROS_MULTIPLIER.toDouble()),
-    val currency: String? = "USD" // TODOBC5 storeProduct?.priceCurrencyCode
+    val storeProduct: StoreProduct? = null, // this appears unused. if we remove, we have to pass full pricing phases
```
[the remark this answers is not in the document]
Comment (swehner) [reply common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt:9]: Would the PurchaseOption object be an option? If not, this seems fine, except the issue with the ids - see comment below

File: feature/google/src/main/java/com/revenuecat/purchases/google/BillingWrapper.kt [hunk 7]
```diff
@@ -217,6 +218,7 @@ class BillingWrapper(
         synchronized(this@BillingWrapper) {
             productTypes[storeProduct.productId] = storeProduct.type
             presentedOfferingsByProductIdentifier[storeProduct.productId] = presentedOfferingIdentifier
+            purchaseOptionSelectedByProductIdentifier[storeProduct.productId] = purchaseOption?.id
```
Comment (swehner) [inline feature/google/src/main/java/com/revenuecat/purchases/google/BillingWrapper.kt:221]: I think purchaseOption can be either baseplan_id or offer_id and they don't have to be unique, so I think we need to change this:
image

File: common/src/main/java/com/revenuecat/purchases/common/Backend.kt (hunk 2 above)
Comment (beylmk) [reply common/src/main/java/com/revenuecat/purchases/common/Backend.kt:196]: oh,  that's good news then, we can remove it :)

File: feature/google/src/main/java/com/revenuecat/purchases/google/BillingWrapper.kt (hunk 7 above)
Comment (beylmk) [reply feature/google/src/main/java/com/revenuecat/purchases/google/BillingWrapper.kt:221]: i'm not sure i understand. our `PurchaseOption.id` will be the base plan ID if the purchase option is simply a base plan, or it will be the offer id if it's an offer. 

i might be misunderstanding what you mean.

Comment (beylmk) [reply feature/google/src/main/java/com/revenuecat/purchases/google/BillingWrapper.kt:221]: ah yes, i see what you mean. i will think about that....

File: public/src/main/java/com/revenuecat/purchases/models/PurchaseOption.kt [hunk 8]
```diff
@@ -6,12 +6,12 @@ import android.os.Parcelable
...
-     * If it represents an offer, it will be the offer ID.
-     * Not applicable for Amazon subscriptions.
-     */
-    val id: String
+
+    // TODO BC5 javadoc
+    val basePlanId: String
+
+    // TODO BC5 javadoc
+    val offerId: String?
```
Comment (swehner) [inline public/src/main/java/com/revenuecat/purchases/models/PurchaseOption.kt:14]: I think this is breaking the abstraction for PurchaseOption, or at least complicating it more. In the end it's tying it again to baseplans/offers.
In this case I'd suggest to just make an idea composed by both (so just `{baseplanId}` or `{baseplanId}:{offerId}`

File: common/src/main/java/com/revenuecat/purchases/common/Backend.kt (hunk 1 above)
Comment (vegaro) [reply common/src/main/java/com/revenuecat/purchases/common/Backend.kt:191]: Yeah, that's tricky. I think we can return a single purchase option with a single princing phase for Amazon? Although it's not really a pricing phase I think it makes for a consistent API

File: public/src/main/java/com/revenuecat/purchases/models/PurchaseOption.kt (hunk 8 above)
Comment (vegaro) [reply public/src/main/java/com/revenuecat/purchases/models/PurchaseOption.kt:14]: Agreed

File: common/src/main/java/com/revenuecat/purchases/common/Backend.kt [hunk 9]
```diff
@@ -189,14 +190,15 @@ class Backend(
...
-            "trial_duration" to receiptInfo.trialDuration,
-            "store_user_id" to storeAppUserID
+            "store_user_id" to storeAppUserID,
+            "pricing_phases" to receiptInfo.pricingPhases?.map { it.toMap() }
         ).filterValues { value -> value != null }
 
-        val extraHeaders = /*receiptInfo.storeProduct?.price?.let { priceString ->
-            mapOf("price_string" to priceString, "marketplace" to marketplace).filterNotNullValues()
-        } ?:*/mapOf<String, String>() // TODOBC5
+        // TODO BC5 we used to only pass this if storeProduct.price was non-null,
```
Comment (beylmk) [inline common/src/main/java/com/revenuecat/purchases/common/Backend.kt:197]: @dalshamaa when you have a moment could you confirm this for me?

File: feature/amazon/src/main/java/com/revenuecat/purchases/amazon/AmazonBilling.kt [hunk 10]
```diff
@@ -439,6 +439,7 @@ internal class AmazonBilling constructor(
             return
         }
 
+        // TODO maddie - could this call normalize purchase data? looks like all it does is get the right id
         amazonBackend.getAmazonReceiptData(
```
Comment (beylmk) [inline feature/amazon/src/main/java/com/revenuecat/purchases/amazon/AmazonBilling.kt:443]: @vegaro is there a reason to not call normalizePurchaseData here?

Comment (beylmk) [reply feature/amazon/src/main/java/com/revenuecat/purchases/amazon/AmazonBilling.kt:443]: again, not related, just came up while i was looking through code

**Comment by beylmk** [comment]
@swehner @vegaro i am getting to the unit tests now, but if you are able to re-review in the meantime that'd be great

File: common/src/main/java/com/revenuecat/purchases/common/Backend.kt (hunk 1 above)
Comment (tonidero) [reply common/src/main/java/com/revenuecat/purchases/common/Backend.kt:192]: Hmm this feels worse to me... Having null values could indicate that we don't have that information, which is correct for amazon. But if we have a single pricing phase for a product with a free trial, that's actually having incorrect information... 

Having said that, if there is no way to detect free trials in amazon and considering this is internal, I think it would be ok. We just need to be aware that pricing phases can contain incomplete information for Amazon. 

Thinking about possible alternatives, we could have different bodies or endpoints in our backend for different stores sending the information we need in a format that matches what we have for each store, but this would be a big change and I don't think it's needed right now so I'm ok with this for now.

**Review by tonidero — COMMENTED** [review]
Nothing major. Haven't reviewed the tests at all until you say they are ready 👍

File: common/src/main/java/com/revenuecat/purchases/common/OfferingFactories.kt [hunk 11]
```diff
@@ -56,8 +56,9 @@ fun JSONObject.createPackage(
     val packageIdentifier = getString("identifier")
     // TODO handle INAPP products
     val productIdentifier = getString("platform_product_identifier")
-    val planIdentifier = getString("platform_product_plan_identifier")
+    val planIdentifier = optString("platform_product_plan_identifier") ?: return null // TODO do we want a crash/error
```
[the remark this answers is not in the document]
Comment (beylmk) [reply common/src/main/java/com/revenuecat/purchases/common/OfferingFactories.kt:59]: oh should've updated this, @vegaro is working on this in a separate PR

File: public/src/main/java/com/revenuecat/purchases/models/PricingPhase.kt [hunk 12]
```diff
@@ -21,4 +21,17 @@ data class PricingPhase(
      * Null for INFINITE_RECURRING or NON_RECURRING recurrence modes.
      */
     val billingCycleCount: Int?
-) : Parcelable
+) : Parcelable {
+
+    fun toMap(): Map<String, Any?> {
+        return mapOf(
+            "billingPeriod" to this.billingPeriod,
+            "billingCycleCount" to (this.billingCycleCount),
```
[the remark this answers is not in the document]
Comment (beylmk) [reply public/src/main/java/com/revenuecat/purchases/models/PricingPhase.kt:29]: oh good catch, accidentally left those in when i was doing a null check

File: public/src/main/java/com/revenuecat/purchases/models/PurchaseOption.kt [hunk 13]
```diff
@@ -8,7 +8,7 @@ import android.os.Parcelable
 interface PurchaseOption : Parcelable {
     /**
      * If this PurchaseOption represents a Google subscription base plan, this will be the base plan ID.
-     * If it represents an offer, it will be the offer ID.
+     * If it represents an offer, it will be {basePlanId}:{offerId}
```
[the remark this answers is not in the document]
Comment (beylmk) [reply public/src/main/java/com/revenuecat/purchases/models/PurchaseOption.kt:11]: that's on the line above, but i should add some clarification

**Review by swehner — APPROVED** [review]
Looks good

File: common/src/test/java/com/revenuecat/purchases/common/BackendTest.kt [hunk 14]
```diff
@@ -74,8 +74,8 @@ class BackendTest {
     private val headersSlot = slot<Map<String, String>>()
```
Comment (beylmk) [inline common/src/test/java/com/revenuecat/purchases/common/BackendTest.kt:74]: these are just formatting updates, i'm going to split unit tests to a separate PR
